pkg/build/api/validation/validation.go

```diff
@@ -149,52 +138,25 @@ func hasProxy(source *buildapi.GitBuildSource) bool {
 	return len(source.HTTPProxy) > 0 || len(source.HTTPSProxy) > 0
 }
 
-func validateSource(input *buildapi.BuildSource) fielderrors.ValidationErrorList {
+func validateSource(input *buildapi.BuildSource, isCustomStrategy bool) fielderrors.ValidationErrorList {
 	allErrs := fielderrors.ValidationErrorList{}
-	switch input.Type {
-	case buildapi.BuildSourceGit:
-		if input.Git == nil {
-			allErrs = append(allErrs, fielderrors.NewFieldRequired("git"))
-		} else {
-			allErrs = append(allErrs, validateGitSource(input.Git).Prefix("git")...)
-		}
-		if input.Dockerfile != nil {
-			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
-		}
-		if input.Binary != nil {
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("binary", "", "may not be set when type is Git"))
-		}
-	case buildapi.BuildSourceBinary:
-		if input.Binary == nil {
-			allErrs = append(allErrs, fielderrors.NewFieldRequired("binary"))
-		} else {
-			allErrs = append(allErrs, validateBinarySource(input.Binary).Prefix("binary")...)
-		}
-		if input.Dockerfile != nil {
-			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
-		}
-		if input.Git != nil {
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("git", "", "may not be set when type is Binary"))
-		}
-	case buildapi.BuildSourceDockerfile:
-		if input.Dockerfile == nil {
-			allErrs = append(allErrs, fielderrors.NewFieldRequired("dockerfile"))
-		} else {
-			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
+
+	if input.Git != nil {
+		// git and binary input sources are mutually exclusive
+		if input.Binary != nil && !isCustomStrategy {
```

Comment:
yes.  they are mutually exclusive.  Either you supply your build content as a binary, or you supply it as a git repository.

that's what the code supports today (if you run start-build --binary the "git" location gets removed), so validation is enforcing what the tooling supports.

and we don't have a use case in which it makes sense to supply both.  binary builds were designed from the outset to be a substitute for git source, not used in conjunction with it.